### PR TITLE
Do not update FabricIndex of PASE session on AddNoC

### PR DESCRIPTION
**Pull request by kghost**
#### Problem
AddNoC updates FabricIndex, but which should be a const.

#### Change overview
Add a effective fabric instead of changing the fabric

#### Testing
Unit tests

### CODE REVIEW

**Comment by mlepage-google** [comment]
This is effectively the same as what I had in the original revision of PR #13712. The problem with that, is it migrates the fabric index for Access Control checks, but since it does not change the accessing fabric index, which is used for filtered reads/writes/etc., those still won't work properly. We're going to need a solution that handles all those cases.

**Comment by mlepage-google** [comment]
The impetus of this PR is to not mess with the peer tuple of (fabric, ID) so that IM/messaging is disturbed (I guess through multi-message interactions?).

How about:
- make the fabric const again
- keep the extra "effective fabric" but change its name to "accessing" fabric
- keep GetFabricIndex returning the const fabric
- make another accessor GetAccessingFabricIndex which returns the new "accessing" fabric if set, else the const fabric
- change GetSubjectDescriptor to use GetAccessingFabricIndex
- change any uses in the code which really want "accessing" fabric to use GetAccessingFabricIndex
- keep any uses in code which need the "stable" const fabric to keep using GetFabricIndex
- document the difference between the two so it's clear when to use which

File: src/transport/SecureSession.cpp [hunk 1]
```diff
@@ -28,13 +28,21 @@ Access::SubjectDescriptor SecureSession::GetSubjectDescriptor() const
         subjectDescriptor.authMode    = Access::AuthMode::kCase;
         subjectDescriptor.subject     = mPeerNodeId;
         subjectDescriptor.cats        = mPeerCATs;
-        subjectDescriptor.fabricIndex = mFabric;
+        VerifyOrDie(mEffectiveFabric == kUndefinedFabricIndex);
```
Comment (tcarmelveilleux) [inline src/transport/SecureSession.cpp:31]: This seems wrong. All CASE subjects should have a subject. It should be:

`VerifyOrDie(mEffectiveFabric != kUndefinedFabricIndex);`

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
But the point is that once this has happened the accessing fabric index should change to be the new fabric index, no?

The fabric of a session is NOT a const in the spec.  Why do we think it should be const, exactly?

**Comment by bzbarsky-apple** [comment]
The plan in https://github.com/project-chip/connectedhomeip/pull/13765#issuecomment-1017693303 makes sense to me.  I would strongly suggest naming the "stable" thing something other than "fabric index"; it sounds like some sort of convenient non-spec-defined stable identifier, and we should figure out what to call it.

Or if it does match some spec concept, which one?

**Comment by woody-apple** [comment]
Fast tracking, given > 3 business days for review.

**Comment by mlepage-google** [comment]
Is that a real test failure?

```
2022-01-20 16:23:59.847 ERROR   16:23:59.775 - TEST OUT  : [1642695839.775867][32210:32215] CHIP:TOO: Warning: currentFabricIndex type checking is not implemented yet. Expected type: 'uint8'
2022-01-20 16:23:59.847 ERROR   16:23:59.775 - TEST OUT  : [1642695839.775928][32210:32215] CHIP:TOO:  ***** Test Failure: currentFabricIndex value < minValue: 0 < 1
```

**Comment by bzbarsky-apple** [comment]
> Is that a real test failure?

Interesting.  That test is reading `currentFabricIndex` and it's getting 0.

But .... I would expect the test to be running CASE, not PASE!  Looking into what's going on there.

**Comment by bzbarsky-apple** [comment]
OK.  So the session is in fact CASE, and has:
```
  mFabric = 1
  mEffectiveFabric = 0
```
but if you look at the PR code, it uses `mEffectiveFabric` for CASE for the subject descriptor, and the subject descriptor is what gets passed to AttributeAccessInterface, so that's what it has to work with.

**Comment by mlepage-google** [comment]
What are we doing to do with this issue, how to resolve? If there's refactoring that needs to occur, what is the plan?

**Pull request closed without merging**